# Netflix Titles — Data Cleaning & Analysis

## 1. Load & First Look

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('netflix_titles.csv')

In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   show_id       8807 non-null   str  
 1   type          8807 non-null   str  
 2   title         8807 non-null   str  
 3   director      6173 non-null   str  
 4   cast          7982 non-null   str  
 5   country       7976 non-null   str  
 6   date_added    8797 non-null   str  
 7   release_year  8807 non-null   int64
 8   rating        8803 non-null   str  
 9   duration      8804 non-null   str  
 10  listed_in     8807 non-null   str  
 11  description   8807 non-null   str  
dtypes: int64(1), str(11)
memory usage: 825.8 KB


In [4]:
df.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

In [5]:
df.describe()

,release_year
count,8807.000000
mean,2014.180198
std,8.819312
min,1925.000000
25%,2013.000000
50%,2017.000000
75%,2019.000000
max,2021.000000


In [6]:
df.head(5)

,show_id,type,title,director,cast,country,date_added,release_year,rating,duration,listed_in,description
0,s1,Movie,Dick Johnson Is Dead,Kirsten Johnson,NaN,United States,"September 25, 2021",2020,PG-13,90 min,Documentaries,"As her father nears the end of his life, filmm..."
1,s2,TV Show,Blood & Water,NaN,"Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...",South Africa,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, TV Dramas, TV Mysteries","After crossing paths at a party, a Cape Town t..."
2,s3,TV Show,Ganglands,Julien Leclercq,"Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...",NaN,"September 24, 2021",2021,TV-MA,1 Season,"Crime TV Shows, International TV Shows, TV Act...",To protect his family from a powerful drug lor...
3,s4,TV Show,Jailbirds New Orleans,NaN,NaN,NaN,"September 24, 2021",2021,TV-MA,1 Season,"Docuseries, Reality TV","Feuds, flirtations and toilet talk go down amo..."
4,s5,TV Show,Kota Factory,NaN,"Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...",India,"September 24, 2021",2021,TV-MA,2 Seasons,"International TV Shows, Romantic TV Shows, TV ...",In a city of coaching centers known to train I...


In [7]:
df.duplicated().unique()

array([False])

## 2. Drop Rows With Small Amounts of Missing Data
`date_added`, `rating`, and `duration` each have only a handful of missing rows (10, 4, and 3 out of 8807) — safe to drop.

In [8]:
df = df[df['date_added'].notnull()]
df = df[df['rating'].notnull()]
df = df[df['duration'].notnull()]
df.isnull().sum()

show_id            0
type               0
title              0
director        2621
cast             825
country          829
date_added         0
release_year       0
rating             0
duration           0
listed_in          0
description        0
dtype: int64

## 3. Explore Unique Values Per Column
`director`, `cast`, and `country` are left with many missing values (2621, 825, 829) — kept as `NaN` rather than dropped, since dropping them would remove a large share of the dataset.

In [9]:
df['show_id'].unique()

<StringArray>
[   's1',    's2',    's3',    's4',    's5',    's6',    's7',    's8',
    's9',   's10',
 ...
 's8798', 's8799', 's8800', 's8801', 's8802', 's8803', 's8804', 's8805',
 's8806', 's8807']
Length: 8790, dtype: str

In [10]:
df['type'].unique()

<StringArray>
['Movie', 'TV Show']
Length: 2, dtype: str

In [11]:
df['title'].unique()

<StringArray>
[            'Dick Johnson Is Dead',                    'Blood & Water',
                        'Ganglands',            'Jailbirds New Orleans',
                     'Kota Factory',                    'Midnight Mass',
 'My Little Pony: A New Generation',                          'Sankofa',
    'The Great British Baking Show',                     'The Starling',
 ...
                        'Zak Storm',                         'Zed Plus',
                            'Zenda',               'Zindagi Gulzar Hai',
                          'Zinzana',                           'Zodiac',
                      'Zombie Dumb',                       'Zombieland',
                             'Zoom',                           'Zubaan']
Length: 8790, dtype: str

In [12]:
df['director'].unique()

<StringArray>
[              'Kirsten Johnson',                             nan,
               'Julien Leclercq',                 'Mike Flanagan',
 'Robert Cullen, José Luis Ucha',                  'Haile Gerima',
               'Andy Devonshire',                'Theodore Melfi',
             'Kongkiat Komesiri',           'Christian Schwochow',
 ...
       'Saratswadee Wongsomphet',             'Kirati Nakintanon',
                   'Mark Risley',                   'James Brown',
                    'Ivona Juka',                        'Mu Chu',
       'Chandra Prakash Dwivedi',               'Majid Al Ansari',
                  'Peter Hewitt',                   'Mozez Singh']
Length: 4527, dtype: str

In [13]:
df['cast'].unique()

<StringArray>
[                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   nan,
                                                                                                                                                                                                                                      'Ama Qamata, Khosi Ngema, Gail Mabalane, Thabang Molaba, Dillon Windvogel, Natasha Thahane, Arno Greeff, Xolile Tshabalala, Getmore Sithole, Cindy Mahlangu, Ryle De Morny, Greteli Fincham, Sello Maake Ka-Ncube, Odwa Gwa

In [14]:
df['country'].unique()

<StringArray>
[                                                        'United States',
                                                          'South Africa',
                                                                     nan,
                                                                 'India',
 'United States, Ghana, Burkina Faso, United Kingdom, Germany, Ethiopia',
                                                        'United Kingdom',
                                               'Germany, Czech Republic',
                                                                'Mexico',
                                                                'Turkey',
                                                             'Australia',
 ...
                        'Germany, United States, United Kingdom, Canada',
          'Canada, India, Thailand, United States, United Arab Emirates',
                             'United States, East Germany, West Germany',
                   

In [15]:
df['date_added'].unique()

<StringArray>
['September 25, 2021', 'September 24, 2021', 'September 23, 2021',
 'September 22, 2021', 'September 21, 2021', 'September 20, 2021',
 'September 19, 2021', 'September 17, 2021', 'September 16, 2021',
 'September 15, 2021',
 ...
    ' March 31, 2017',     'March 23, 2017',  'December 25, 2016',
  'November 30, 2016',   'October 23, 2017',   'November 4, 2017',
     'August 5, 2015',   'December 6, 2018',      'March 9, 2016',
   'January 11, 2020']
Length: 1765, dtype: str

In [16]:
df['release_year'].unique()

array([2020, 2021, 1993, 2018, 1996, 1998, 1997, 2010, 2013, 2017, 1975,
       1978, 1983, 1987, 2012, 2001, 2014, 2002, 2003, 2004, 2011, 2008,
       2009, 2007, 2005, 2006, 1994, 2015, 2019, 2016, 1982, 1989, 1990,
       1991, 1999, 1986, 1992, 1984, 1980, 1961, 2000, 1995, 1985, 1976,
       1959, 1988, 1981, 1972, 1964, 1945, 1954, 1979, 1958, 1956, 1963,
       1970, 1973, 1925, 1974, 1960, 1966, 1971, 1962, 1969, 1977, 1967,
       1968, 1965, 1946, 1942, 1955, 1944, 1947, 1943])

In [17]:
df['rating'].unique()

<StringArray>
[   'PG-13',    'TV-MA',       'PG',    'TV-14',    'TV-PG',     'TV-Y',
    'TV-Y7',        'R',     'TV-G',        'G',    'NC-17',       'NR',
 'TV-Y7-FV',       'UR']
Length: 14, dtype: str

In [18]:
df['duration'].unique()

<StringArray>
[   '90 min', '2 Seasons',  '1 Season',    '91 min',   '125 min', '9 Seasons',
   '104 min',   '127 min', '4 Seasons',    '67 min',
 ...
    '43 min',   '200 min',   '196 min',   '167 min',   '178 min',   '228 min',
    '18 min',   '205 min',   '201 min',   '191 min']
Length: 220, dtype: str

In [19]:
df['listed_in'].unique()

<StringArray>
[                                                'Documentaries',
               'International TV Shows, TV Dramas, TV Mysteries',
 'Crime TV Shows, International TV Shows, TV Action & Adventure',
                                        'Docuseries, Reality TV',
        'International TV Shows, Romantic TV Shows, TV Comedies',
                            'TV Dramas, TV Horror, TV Mysteries',
                                      'Children & Family Movies',
              'Dramas, Independent Movies, International Movies',
                                  'British TV Shows, Reality TV',
                                              'Comedies, Dramas',
 ...
                               'Comedies, Dramas, Sports Movies',
         'Action & Adventure, Romantic Movies, Sci-Fi & Fantasy',
                        'Classic & Cult TV, TV Sci-Fi & Fantasy',
                           'Comedies, Cult Movies, LGBTQ Movies',
                     'Comedies, Horror Movies, Sci-Fi & F

## 4. Clean `date_added`
Some values had a leading space (e.g. `' March 31, 2017'`); `.str.strip()` removes it before converting to a real date type.

In [20]:
df['date_added'] = df['date_added'].str.strip()
df['date_added'] = pd.to_datetime(df['date_added'])
df['date_added'].unique()

<DatetimeArray>
['2021-09-25 00:00:00', '2021-09-24 00:00:00', '2021-09-23 00:00:00',
 '2021-09-22 00:00:00', '2021-09-21 00:00:00', '2021-09-20 00:00:00',
 '2021-09-19 00:00:00', '2021-09-17 00:00:00', '2021-09-16 00:00:00',
 '2021-09-15 00:00:00',
 ...
 '2018-09-27 00:00:00', '2017-03-23 00:00:00', '2016-12-25 00:00:00',
 '2016-11-30 00:00:00', '2017-10-23 00:00:00', '2017-11-04 00:00:00',
 '2015-08-05 00:00:00', '2018-12-06 00:00:00', '2016-03-09 00:00:00',
 '2020-01-11 00:00:00']
Length: 1713, dtype: datetime64[us]

---
# Questions

### Q1 — What is the ratio of Movies to TV Shows?

In [21]:
df['type'].value_counts(normalize=True) * 100

type
Movie      69.692833
TV Show    30.307167
Name: proportion, dtype: float64

### Q2 — Which country has produced the most content?
`country` holds multiple comma-separated values per row, so each row is split and exploded before counting.

In [22]:
country_list = (
    df['country'].str.split(', ')
    .explode()
    .value_counts()
    .head(10)
)
country_list

country
United States     3680
India             1046
United Kingdom     803
Canada             445
France             393
Japan              316
Spain              232
South Korea        231
Germany            226
Mexico             169
Name: count, dtype: int64

### Q3 — Which genre (`listed_in`) is the most common?

In [23]:
genre_count = (
    df['listed_in'].str.split(', ')
    .explode()
    .value_counts()
    .head(10)
)
genre_count

listed_in
International Movies        2752
Dramas                      2426
Comedies                    1674
International TV Shows      1349
Documentaries                869
Action & Adventure           859
TV Dramas                    762
Independent Movies           756
Children & Family Movies     641
Romantic Movies              616
Name: count, dtype: int64

### Q4 — What is the distribution of `rating` (content age rating)?

In [24]:
df['rating'].value_counts(normalize=True) * 100

rating
TV-MA       36.461889
TV-14       24.539249
TV-PG        9.795222
R            9.089875
PG-13        5.574516
TV-Y7        3.788396
TV-Y         3.481229
PG           3.265074
TV-G         2.502844
NR           0.898749
G            0.466439
TV-Y7-FV     0.068259
NC-17        0.034130
UR           0.034130
Name: proportion, dtype: float64

### Q5 — Which director has made the most titles?

In [25]:
df['director'].str.split(', ').explode().value_counts().head(10)

director
Rajiv Chilaka          22
Jan Suter              21
Raúl Campos            19
Suhas Kadav            16
Marcus Raboy           16
Jay Karas              15
Cathy Garcia-Molina    13
Youssef Chahine        12
Martin Scorsese        12
Jay Chapman            12
Name: count, dtype: int64

### Q6 — Which actor appears in the most titles?

In [26]:
df['cast'].str.split(', ').explode().value_counts().head(10)

cast
Anupam Kher         43
Shah Rukh Khan      35
Julie Tejwani       33
Naseeruddin Shah    32
Takahiro Sakurai    32
Rupa Bhimani        31
Akshay Kumar        30
Om Puri             30
Yuki Kaji           29
Amitabh Bachchan    28
Name: count, dtype: int64

### Q7 — How has the number of titles added per year changed?
Sorted by year (`sort_index`), not by count, so the trend over time is actually visible.

Steady growth from 2008 to 2019. The drop in 2020 lines up with COVID-related production slowdowns. The drop in 2021 is not a real decline — this dataset was collected around September 2021, so that year is only ~9 months complete.

In [27]:
df['date_added'].dt.year.value_counts().sort_index()

date_added
2008       2
2009       2
2010       1
2011      13
2012       3
2013      11
2014      24
2015      82
2016     426
2017    1185
2018    1648
2019    2016
2020    1879
2021    1498
Name: count, dtype: int64

### Q8 — Which `release_year` has the most titles?

In [28]:
df.groupby('release_year').size().sort_values(ascending=False).head(10)

release_year
2018    1146
2019    1030
2017    1030
2020     953
2016     901
2021     592
2015     555
2014     352
2013     286
2012     236
dtype: int64

Breakdown by type as well — 2018 leads for Movies specifically too:

In [29]:
df.groupby(['release_year', 'type']).size().sort_values(ascending=False).head(10)

release_year  type   
2018          Movie      767
2017          Movie      765
2016          Movie      658
2019          Movie      633
2020          Movie      517
              TV Show    436
2019          TV Show    397
2015          Movie      396
2018          TV Show    379
2021          TV Show    315
dtype: int64

### Q9 — Has the Movie/TV Show ratio changed over the years?

In [30]:
df.groupby('release_year')['type'].value_counts(normalize=True).unstack() * 100

type,Movie,TV Show
release_year,,
1925,NaN,100.000000
1942,100.000000,NaN
1943,100.000000,NaN
1944,100.000000,NaN
1945,75.000000,25.000000
...,...,...
2017,74.271845,25.728155
2018,66.928447,33.071553
2019,61.456311,38.543689


### Q10 — Do different countries focus on different genres?
Both `country` and `listed_in` are multi-value columns, so both are split and exploded (independently, not paired) to get every (country, genre) combination.

`.assign()` is used instead of overwriting `df['country']` directly — this keeps the original `df` untouched, so this cell is safe to re-run without corrupting anything.

In [31]:
df_exploded = (
    df.assign(
        country=df['country'].str.split(', '),
        listed_in=df['listed_in'].str.split(', ')
    )
    .explode('country')
    .explode('listed_in')
)
df_exploded = df_exploded[df_exploded['country'] != '']

country_genre = df_exploded.groupby(['country', 'listed_in']).size()
top_genre_per_country = country_genre.groupby(level=0).idxmax().str[1]
top_genre_per_country

country
Afghanistan            Documentaries
Albania                       Dramas
Algeria                       Dramas
Angola            Action & Adventure
Argentina       International Movies
                        ...         
Vatican City           Documentaries
Venezuela              Documentaries
Vietnam         International Movies
West Germany    International Movies
Zimbabwe        International Movies
Length: 126, dtype: object